### PR TITLE
ES6: migrated "entity" (part 2)

### PR DESCRIPTION
**Pull request by bennyn**

### CODE REVIEW

File: app/script/entity/Conversation.js [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.entity = z.entity || {};
+
+
```
Comment (gregor) [inline app/script/entity/Conversation.js]: One too many

File: app/script/entity/Conversation.js [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+   * @param {string} conversation_id - Conversation ID
+   */
+  constructor(conversation_id = '') {
+    this.id = conversation_id;
+    this.creator = undefined;
+    this.type = ko.observable();
+    this.name = ko.observable();
+    this.input = ko.observable(z.util.StorageUtil.get_value(`${z.storage.StorageKey.CONVERSATION.INPUT}|${this.id}`) || '');
+    this.input.subscribe((text) => {
+      return z.util.StorageUtil.set_value(`${z.storage.StorageKey.CONVERSATION.INPUT}|${this.id}`, text);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: I guess we can keep simply things like this in a single line

File: app/script/entity/Conversation.js [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+    });
+
+    this.is_pending = ko.observable(false);
+    this.is_loaded = ko.observable(false);
+
+    this.participating_user_ets = ko.observableArray([]); // Does not include us
+    this.participating_user_ids = ko.observableArray([]);
+    this.self = undefined;
+    this.number_of_participants = ko.pureComputed(() => {
+      return this.participating_user_ids().length;
```
Comment (gregor) [inline app/script/entity/Conversation.js]: See above

File: app/script/entity/Conversation.js [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+     * - Join the user's first names to a comma separated list or uses the user's first name if only one user participating
+     * - "..." if the user entities have not yet been attached yet
+     */
+    this.display_name = ko.pureComputed(() => {
+      if ([z.conversation.ConversationType.CONNECT, z.conversation.ConversationType.ONE2ONE].includes(this.type())) {
+        if (this.participating_user_ets()[0] && this.participating_user_ets()[0].name) {
+          return this.participating_user_ets()[0].name();
+        }
+        return '…';
+      } else if (this.is_group()) {
```
Comment (gregor) [inline app/script/entity/Conversation.js]: No need for else. You always return in the above if

File: app/script/entity/Conversation.js [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+
+  /**
+   * Adds a single message to the conversation.
+   * @param {z.entity.Message} message_et - Message entity to be added to the conversation
+   * @returns {undefined} No return value
+   */
+  add_message(message_et) {
+    amplify.publish(z.event.WebApp.CONVERSATION.MESSAGE.ADDED, message_et);
+    this._update_last_read_from_message(message_et);
+    return this.messages_unordered.push(this._check_for_duplicate_nonce(message_et, this.get_last_message()));
```
Comment (gregor) [inline app/script/entity/Conversation.js]: Don't return if you return undefined

File: app/script/entity/Conversation.js [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+    // we found a message from self user
+    for (let counter = message_ets.length - 1; counter >= 0; counter--) {
+      message_et = message_ets[counter];
+      if (message_et.user() && message_et.user().is_me) {
+        this._update_last_read_from_message(message_et);
+        break;
+      }
+    }
+
+    return z.util.ko_array_push_all(this.messages_unordered, message_ets);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: Do we really want to return this?

File: app/script/entity/Conversation.js [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+  prepend_messages(message_ets) {
+    const last_message_et = message_ets[message_ets.length - 1];
+    this._check_for_duplicate_nonce(last_message_et, this.get_first_message());
+
+    for (let index = message_ets.length - 1; index >= 0; index--) {
+      let message_et = message_ets[index];
+      message_et = this._check_for_duplicate_nonce(message_ets[index - 1], message_et);
+    }
+
+    return z.util.ko_array_unshift_all(this.messages_unordered, message_ets);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: Return?

File: app/script/entity/Conversation.js [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+    }
+  }
+
+  /**
+   * Removes a single message from the conversation.
+   * @param {z.entity.Message} message_et - Message entity to be removed from the conversation
+   * @returns {undefined} No return value
+   */
+  remove_message(message_et) {
+    return this.messages_unordered.remove(message_et);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: Sure we want to return this?

File: app/script/entity/Conversation.js [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+  }
+
+  /**
+   * Removes all messages from the conversation.
+   * @returns {undefined} No return value
+   */
+  remove_messages() {
+    return this.messages_unordered.removeAll();
```
Comment (gregor) [inline app/script/entity/Conversation.js]: See above

File: app/script/entity/Conversation.js [hunk 10]
```diff
@@ -0,0 +1,628 @@
...
+  }
+
+  /**
+   * Checks for message duplicates by nonce and returns the message.
+   * @private
+   * @note If a message is send to the backend multiple times by a client they will be in the conversation multiple times
+   * @param {z.entity.Message} message_et - Message entity to be added to the conversation
+   * @param {z.entity.Message} other_message_et - Other message entity to compare with
+   * @returns {undefined} No return value
```
Comment (gregor) [inline app/script/entity/Conversation.js:385]: You return a message entity

File: app/script/entity/Conversation.js [hunk 11]
```diff
@@ -0,0 +1,628 @@
...
+    }
+
+    return message_et;
+  }
+
+  /**
+   * Creates the placeholder message after clearing a conversation.
+   * @private
+   * @note Only create the message if the group participants have been set
+   * @returns {undefined} No return value
```
Comment (gregor) [inline app/script/entity/Conversation.js:411]: You return a message entity

File: app/script/entity/Conversation.js [hunk 12]
```diff
@@ -0,0 +1,628 @@
...
+  }
+
+  /**
+   * Update information about last activity from single message.
+   * @param {z.entity.Message} message_et - Message to be added to conversation
+   * @returns {undefined} No return value
+   */
+  update_latest_from_message(message_et) {
+    if ((message_et != null) && message_et.visible() && message_et.should_effect_conversation_timestamp) {
+      return this.set_timestamp(message_et.timestamp(), z.conversation.ConversationUpdateType.LAST_EVENT_TIMESTAMP);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: Return?

File: app/script/entity/Conversation.js [hunk 13]
```diff
@@ -0,0 +1,628 @@
...
+   * @private
+   * @param {z.entity.Message} message_et - Message entity
+   * @returns {undefined} No return value
+   */
+  _update_last_read_from_message(message_et) {
+    const is_me = message_et.user() && message_et.user().is_me;
+    const has_timestamp = message_et.timestamp();
+
+    if (is_me && has_timestamp && message_et.should_effect_conversation_timestamp) {
+      return this.set_timestamp(message_et.timestamp(), z.conversation.ConversationUpdateType.LAST_READ_TIMESTAMP);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: Return?

File: app/script/entity/Conversation.js [hunk 14]
```diff
@@ -0,0 +1,628 @@
...
+    this.archived_timestamp = ko.observable(0);
+    this.cleared_timestamp = ko.observable(0);
+    this.last_event_timestamp = ko.observable(0);
+    this.last_read_timestamp = ko.observable(0);
+    this.muted_timestamp = ko.observable(0);
+
+    // Conversation states for view
+    this.is_muted = ko.pureComputed(() => {
+      return this.muted_state();
+    });
```
Comment (ffflorian) [inline app/script/entity/Conversation.js]: I'd suggest
```js
this.is_muted = ko.pureComputed(() => this.muted_state());
```
everywhere where functions consist only of one `return` line.

File: app/script/entity/Conversation.js [hunk 15]
```diff
@@ -0,0 +1,628 @@
...
+    }
+
+    return pending_uploads.length;
+  }
+
+  get_users_with_unverified_clients() {
+    return [this.self]
+      .concat(this.participating_user_ets())
+      .filter((user_et) => !user_et.is_verified())
+      .map((user_et) => user_et);
```
Comment (gregor) [inline app/script/entity/Conversation.js]: No need for this map

File: app/script/entity/Conversation.js (hunk 2 above)
Comment (bennycode) [reply app/script/entity/Conversation.js]: It extended our character limit per line:

unbenannt

File: app/script/entity/Conversation.js (hunk 11 above)
Comment (bennycode) [reply app/script/entity/Conversation.js:411]: Not anymore 🗡

File: app/script/entity/Conversation.js (hunk 10 above)
Comment (bennycode) [reply app/script/entity/Conversation.js:385]: Not anymore 🗡

**Comment by bennycode** [comment]
@gregor & @ffflorian: I fixed all of your requests.

File: app/script/entity/Conversation.js [hunk 16]
```diff
@@ -0,0 +1,623 @@
...
+      this.last_read_timestamp,
+      this.muted_state,
+      this.name,
+      this.participating_user_ids,
+      this.status,
+      this.type,
+      this.verification_state,
+    ].forEach((property) => {
+      return property.subscribe(this.persist_state);
+    });
```
Comment (ffflorian) [inline app/script/entity/Conversation.js]:
```js
].forEach((property) => property.subscribe(this.persist_state));
```

File: app/script/entity/User.js [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+      }
+    }, this, {deferEvaluation: true});
+
+    this.email = ko.observable();
+    this.phone = ko.observable();
+
+    this.name = ko.observable('');
+    this.first_name = ko.pureComputed(() => {
+      return this.name().split(' ')[0];
+    });
```
Comment (ffflorian) [inline app/script/entity/User.js]:
```js
this.first_name = ko.pureComputed(() => this.name().split(' ')[0]);
```

**Pull request merged**